tests/test_main.py

```diff
@@ -0,0 +1,318 @@
+from click.testing import CliRunner
+from moodlecli.main import cli
+
+TEST_MOODLE_URL = "http://localhost:8000"
+TEST_MOODLE_TOKEN = "94f9362009170815d912b149bc1e5ffa"
+
+MOODLE_WEBSERVICE_PATH = "/webservice/rest/server.php"
+
+
+def get_matching_helper(request, context):
+    wsfunction = request.qs["wsfunction"][0]
+
+    if wsfunction == moodle.MOODLE_FUNC_GET_ROLE_BY_SHORTNAME:
+        return {'id': 2}
+    elif wsfunction == moodle.MOODLE_FUNC_GET_USERS:
+        # Used to exercise error in course_bulk_setup
+        if request.qs["criteria[0][value]"][0] == 'invalid@gmail.com':
+            return {'exception': 'value'}
+        else:
+            return {'users': [{'id': 2}]}
+    elif wsfunction == moodle.MOODLE_FUNC_GET_SELF_ENROLMENT_METHODS:
+        return [{'id': 2}]
+    else:
+        return []
+
+
+def post_matching_helper(request, context):
+
+    wsfunction = parse.parse_qs(request.body)['wsfunction'][0]
+
+    if wsfunction == moodle.MOODLE_FUNC_DUPLICATE_COURSE:
+        return {'id': 3}
+    elif wsfunction == moodle.MOODLE_FUNC_ENABLE_SELF_ENROLMENT_METHOD:
+        return {'id': 2}
+    elif wsfunction == moodle.MOODLE_FUNC_CREATE_USERS:
+        return {'users': {'id': 5}}
+    elif wsfunction == moodle.MOODLE_FUNC_IMPORT_COURSE:
+        return {'id': 3}
+    elif wsfunction == moodle.MOODLE_FUNC_SET_SELF_ENROLMENT_METHOD_KEY:
+        return {'id': 3}
+    else:
+        return []
+
+
+def test_copy_course(requests_mock):
+    runner = CliRunner()
+    requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                       json=post_matching_helper)
+
+    # copy-course <SOURCE_ID> <COURSE_NAME> <COURSE_SHORT_NAME>
+    # <COURSE_CATEGORY_ID>
+    result = runner.invoke(cli, ['copy-course', '3', 'Algebra3', 'ALG3', '1'],
+                           env={'MOODLE_URL': TEST_MOODLE_URL,
+                                'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+    assert result.exit_code == 0
+
+
+def test_course_bulk_csv(requests_mock, tmp_path):
+    test_json = {'foo': 'bar'}
+    runner = CliRunner()
+    with runner.isolated_filesystem(temp_dir=tmp_path):
+        requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                          json=test_json)
+        # course-bulk-csv  <OUTPUT_CSV>
+        result = runner.invoke(cli, ['course-bulk-csv', 'output.csv'],
+                               env={'MOODLE_URL': TEST_MOODLE_URL,
+                                    'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+        assert result.exit_code == 0
+        assert os.stat("output.csv").st_size != 0
+
+
+def test_course_bulk_setup(requests_mock, tmp_path):
+    runner = CliRunner()
+    with runner.isolated_filesystem(temp_dir=tmp_path):
+        requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                          json=get_matching_helper)
+        requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                           json=post_matching_helper)
```

Comment:
Just a thought on if you want to DRY these up: One thing you can do is create a [custom fixture](https://docs.pytest.org/en/latest/reference/fixtures.html) in this file something like:

```python
@pytest.fixture
def moodle_requests_mock(requests_mock):
    requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}', json=get_matching_helper)
    requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}', json=post_matching_helper)
```

Then, in every test that needs the mock setup, just pass `moodle_requests_mock` instead of `requests_mock` and you'll get your configured mock automatically:

```python
def test_course_bulk_setup(moodle_requests_mock, tmp_path):
```

It probably doesn't even matter that some tests only need the `get` mock...you can use the `moodle_requests_mock` in all of them to keep things simple.

Another advantage to :point_up: is you can shove the helpers into the fixture function since they'd no longer need to be in the global scope.

Reply:
This sounds great. Would the other helper functions be automatically included in this scope or should I add a `@pytest.fixture` to those functions as well?